Add misc utils

Roynecro97 opened the pull request:
Add various utilities for handling (cyclic) value ranges and creating strong types.

* `StronglyTyped` - A base class for creating strongly typed primitives (and other types).

  Usage:
  ```cpp
  using MyStrongInt = StronglyTyped<int>;
  ```

* `ValueRange` - A helper for handling `clamp`/`fix_cycle` (modulo arithmetic) operations on values in a range.
* `ConstrainedValue` - A wrapper for variables that have values within a value range.
* `AngleRange` - A `ValueRange` of `[-pi, pi)`.
* `Angle` - A `ConstrainedValue` with `AngleRange`.
* Physical unit handling:
  * `PhysicalSize` - A class template representing physical sizes (distance, time, velocity, etc.).
  * `unit_cast` - Converts a `PhysicalSize` to a different inner representation or ratio.

Roynecro97 committed:
Add initial direction enums for the maze solver component

Also add the component itself.

Roynecro97 committed:
Add `misc_utils` component

Initial utils:
* StronglyTyped
* Type utilities:
  * PartialArithmetic - A concept for a type that supports basic
    arithmetic operations.
  * Arithmetic - std::integer || std::floating_point.
  * FloatingPointWrapper - A PartialArithmetic type that eventually
    wraps a floating point type. Meaning there is a chain of
    T::type::type::...::type where the last `type` is a
    std::floating_point and all types are PartialArithmetics.
  * ExtendedFloatingPoint - A floating point primitive type or a
    FloatingPointWrapper.
  * make_floating_point - Gets the underlying floating point type for an
    ExtendedFloatingPoint type.
* ValueRange (AngleRange)
* ConstrainedValue (Angle)

Roynecro97 committed:
Simplify subtraction wraparound logic

Roynecro97 committed:
Upgrade `to_radians(Direction)` to work for any `ExtendedFloatingPoint`

OmerHuly commented:
Add `sin` and `cos` functions for `Angle` please.

OmerHuly commented on components/maze_solver/include/maze_solver/direction.h:
```diff
@@ -0,0 +1,237 @@
+#ifndef MAZE_SOLVER_DIRECTION_H
+#define MAZE_SOLVER_DIRECTION_H
+
+#include "misc_utils/type_utils.h"
```
```suggestion
#include <misc_utils/typing_utils.h>
```

Roynecro97 committed:
Change `StronglyTyped` to use automatic tagging instead of CRTP

Also change `struct` -> `class`.

Usage changes from:

```cpp
struct MyInt : StronglyTyped<int, MyInt>
{
    using StronglyTyped<int, MyStrongInt>::StronglyTyped;
};
```

To a simple type alias:

```cpp
using MyInt = StronglyTyped<int>;
```

Roynecro97 committed:
Add `ValueRange<...>::contains()`

Roynecro97 committed:
Fix `ConstrainedValue`'s behavior with special ranges

* Fix wrong handling of the default constructor when the range doesn't
  contain `T{}`.
* `struct` -> `class` and make the value member private to avoid
  external modifications.
* Implicit conversion to `T` instead of explicit for easier use with
  standard math functions. For an explicit conversion, use a
  `StronglyTyped<ConstrainedValue<...>>`.

Roynecro97 committed:
Add utilities for handling physical units

OmerHuly commented on components/misc_utils/include/misc_utils/strongly_typed.h:
```diff
@@ -0,0 +1,335 @@
+    friend constexpr StronglyTyped operator*(const StronglyTyped &lhs, const StronglyTyped &rhs)
+        noexcept(noexcept(lhs.value * rhs.value))
+        requires (requires (T a, T b) {
+            { a *b } -> std::convertible_to<type>;
```
Clang format 😢

OmerHuly commented on components/misc_utils/include/misc_utils/strongly_typed.h:
```diff
@@ -0,0 +1,335 @@
+    friend constexpr StronglyTyped operator&(const StronglyTyped &lhs, const StronglyTyped &rhs)
+        noexcept(noexcept(lhs.value & rhs.value))
+        requires (requires (T a, T b) {
+            { a &b } -> std::convertible_to<type>;
```
Clang format 😢

OmerHuly commented on components/misc_utils/include/misc_utils/physical_size.h:
```diff
@@ -0,0 +1,366 @@
+
+    template <PartialArithmetic RepU, RatioSpec RatioU>
+        requires (requires (Rep a, RepU b) {
+            { a *b } -> std::convertible_to<Rep>;
```
Clang format 😢

OmerHuly commented on components/misc_utils/include/misc_utils/physical_size.h:
```diff
@@ -0,0 +1,366 @@
+
+    template <std::convertible_to<Rep> T>
+        requires requires (Rep a, T b) {
+            { a *b } -> std::convertible_to<Rep>;
```
Clang format 😢

OmerHuly commented on components/misc_utils/include/misc_utils/physical_size.h:
```diff
@@ -0,0 +1,366 @@
+
+    template <std::convertible_to<Rep> T>
+        requires requires (T a, Rep b) {
+            { a *b } -> std::convertible_to<Rep>;
```
Clang format 😢

OmerHuly commented on components/misc_utils/include/misc_utils/unit_symbols.h:
```diff
@@ -0,0 +1,369 @@
+    static constexpr auto den_sym = std::string_view(detail::symbol_helper<typename T::den>::symbol);
+
+    static consteval auto get_symbol_impl()
+
```
Clang format 😢

Roynecro97 commented:
> Add `sin` and `cos` functions for `Angle` please.

Update: we eventually decided on making `ConstrainedValue`'s (`Angle`'s) conversion operator implicit and require a `StronglyTyped<...>` wrapper for an explicit cast.

Roynecro97 committed:
Make `PhysicalSize` use it's type aliases instead of the template params

Unify style with other class templates in the code.

Roynecro97 committed:
Fix `PhysicalSize` conversions

* Fix constraints on the conversion / in-place constructor.
* Fix `noexcept` specifiers.
* Add an implicit conversion for `PhysicalSize`s with no units and a 1:1
  ratio.
* Add an explicit conversion for `PhysicalSize`s with no units and a
  custom ratio.

Roynecro97 committed:
Add a type alias for [m/s] `PhysicalSize`

Roynecro97 committed:
Add `unit_mul`/`unit_div` specializations with `PhysicalUnit`s

Roynecro97 committed:
Refactor `get_symbol` implementation to have a unified implementation

Roynecro97 committed:
Simplify `PhysicalUnit` to avoid `TString` and lambda tagging

Lambda tagging can cause linker problems, it was redundant here.

`TString` works well, but caused problems with the linter. This form
doesn't cause errors in the linter.

Roynecro97 committed:
Revert "Change `StronglyTyped` to use automatic tagging instead of CRTP"

This reverts commit 77b83de90885d09bbaf8e53af957ebd45a4e90aa.

Automatic tagging is undesired if the strong type alias is declared in a
header as it can cause linkage problems.

Roynecro97 committed:
Add support for both CRTP / Tag modes in StronglyTyped

Also add pointer-like API (similar to `std::optional`).

New usages are:

```cpp
struct Int : StronglyTypedBase<int, Int>
{
    using StronglyTypedBase<int, Int>::StronglyTypedBase;
};
```

or

```cpp
struct int_tag;
using Int = StronglyTyped<int, int_tag>;
```

or (for types defined only in a .cpp file)

```cpp
using Int = StronglyTyped<int>;
```

Roynecro97 committed:
Add deduction guides for `PhysicalSize`

Roynecro97 committed:
Fix return value consistency for directions

Roynecro97 committed:
Fix StronglyTyped shift operators' cyclic definition

Roynecro97 committed:
Add enum2str for ValueRange Mode

Roynecro97 committed:
Add initial unit tests

OmerHuly commented on main/unittests/strongly_typed_test.cc:
```diff
@@ -0,0 +1,238 @@
+// static_assert(!std::is_convertible_v<SimpleStrongInt, TaggedStrongInt>::value);
+// static_assert(!std::is_convertible_v<TaggedStrongInt, SimpleStrongInt>::value);
+// static_assert(!std::is_convertible_v<VeryStrongInt, TaggedStrongInt>::value);
+// static_assert(!std::is_convertible_v<TaggedStrongInt, VeryStrongInt>::value);
```
Do you want to leave those comments?
